server: do not report `["encountered error"] [error=EOF]` for TCPConn alive detectors

## Enhancement

After https://github.com/pingcap/tidb/pull/27962, TiDB reports the handshake error if the connection is not successfully established, as a trivial bugfix. This change is released with v5.3.0

However, we found that some load balancers use the TCP connection for alive detectives, for example, the [GCP](https://cloud.google.com/load-balancing/docs/network) and https://asktug.com/t/topic/273508.

We can easily reproduce the following error logs from TiDB:

```
#!/usr/bin/env python    
    
import socket    
    
TCP_IP = '127.0.0.1'    
TCP_PORT = 4000    
BUFFER_SIZE = 1024    
MESSAGE = ""    
    
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)    
s.connect((TCP_IP, TCP_PORT))    
s.close()    
```

You can see the following error logs from TiDB:
```
[ERROR] [terror.go:307] ["encountered error"] [error=EOF] [stack="github.com/pingcap/tidb/parser/terror.Log\n\t/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/parser/terror/terror.go:307\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/server.go:519"]
```

The although harmless, the logs can be confusing to the users, so I think it's better to change the log level to 'DEBUG' when the `io.EOF` error is encountered: https://github.com/pingcap/tidb/blob/abb65829101d2f19b69b4e5a4dd0471dbed05915/server/server.go#L513

BTW, MySQL reports the following logs for the case above:
```
2021-12-27T14:03:16.793444Z 12 [Note] Got an error reading communication packets
```
So the failed connection logs happened for MySQL, but it is better in `Note` level.

--- a/server/server.go
+++ b/server/server.go
@@ -33,6 +33,7 @@ import (
 	"context"
 	"crypto/tls"
 	"fmt"
+	"io"
 	"math/rand"
 	"net"
 	"net/http"
@@ -510,11 +511,17 @@ func (s *Server) onConn(conn *clientConn) {
 			})
 			terror.Log(err)
 		}
-		// Some keep alive services will send request to TiDB and disconnect immediately.
-		// So we only record metrics.
-		metrics.HandShakeErrorCounter.Inc()
-		terror.Log(errors.Trace(err))
-		terror.Log(errors.Trace(conn.Close()))
+		if errors.Cause(err) == io.EOF {
+			// `EOF` means the connection is closed normally, we do not treat it as a noticeable error and log it in 'DEBUG' level.
+			logutil.BgLogger().With(zap.Uint64("conn", conn.connectionID)).
+				Debug("EOF", zap.String("remote addr", conn.bufReadConn.RemoteAddr().String()))
+		} else {
+			metrics.HandShakeErrorCounter.Inc()
+			logutil.BgLogger().With(zap.Uint64("conn", conn.connectionID)).
+				Warn("Server.onConn handshake", zap.Error(err),
+					zap.String("remote addr", conn.bufReadConn.RemoteAddr().String()))
+		}
+		terror.Log(conn.Close())
 		return
 	}
 

